### PR TITLE
Complete provenance recording

### PR DESCRIPTION
**Pull request by benjeffery**
Issue #912 

I've started looking into this, the following `simulate` arguments use non-basic types that would need serialising:
- [x] recombination_map (`RecombinationMap`)
- [x] population_configurations (list of `PopulationConfiguration` or None.)
- [x] migration_matrix (list or numpy array)
- [x] demographic_events (list of subclasses of `DemographicEvent`)
- [x] samples (list of `Sample`)
- [x] from_ts (`TreeSequence`) 
- [x] model (string or `SimulationModel`)

Also the following two undocumented arguments have non-basic types:
- [x] pedigree (`Pedigree`)
- [x] ~~mutation_generator (`_msprime.MutationGenerator`)~~ Deprecated since 0.6.1 so don't include

Some of these classes do not support serialising by inspecting by `cls.__dict__`, for example `RecombinationMap` doesn't store it's arguments but passes them to the low-level instance. ~~My plan is to add `to_json` methods to these classes.~~ `asdict` methods as this allows easier recursive serialisation.

- [x] consider security implications of functional serialisation
- [x]  log errors in final provenance
- [x] `record_provenance` option to disable writing
- [x] `parse_simulate_provenance` to give dict of args from a provenance
- [x] replication index issues
- [x] `mutate` provenance
- [x] `parse_mutate_provenance`

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Sounds like a plan @benjeffery. Quick thoughts:

- For ``from_ts`` I think we should use the ``TreeSequence.uuid`` property which I *think* is always defined. I guess the only other thing we could do is use the ``TreeSequence.asdict()`` method and coerce that into JSON. 

- We should be able to convert the ``Pedigree`` object to JSON easily enough (but see #850)

**Comment by benjeffery** [comment]
Making progress.
Two questions I'm working on:

- Dealing with functional parameters such as those at `test_models.py:614`. I'm tempted to serialise those, you can't do this with `pickle` but you can with `dill`, as long as they aren't higher-order recursive functions. This would mean bringing in another dependency unless I extract the function serialiser from `dill` .

- For tree sequences are you talking about `file_uuid`? It seems to be `None`:
```
In [7]: msprime.simulate(10).file_uuid == None                                                                                                                                                                                                
Out[7]: True
```

**Comment by jeromekelleher** [comment]
> Dealing with functional parameters such as those at test_models.py:614. I'm tempted to serialise those, you can't do this with pickle but you can with dill, as long as they aren't higher-order recursive functions. This would mean bringing in another dependency unless I extract the function serialiser from dill

Yuck, hadn't thought of that.  Not sure what to do there, possibly just print out a warning saying "incomplete provenance or something". I don't think this can be solved in general.

> For tree sequences are you(@jeromekelleher) talking about file_uuid? It seems to be None:

Yes I was. Oh well, looks like it's an attribute of the file so not stored here.

**Comment by benjeffery** [comment]
> Yuck, hadn't thought of that. Not sure what to do there, possibly just print out a warning saying "incomplete provenance or something". I don't think this can be solved in general.

As a compromise I'm only serialising functions that don't refer to globals or the outer scope for now, although it would be possible to do more.

**Comment by benjeffery** [comment]
Provenance can now be round-tripped successfully for `simulate`, rudimentary tests for this are in place, will add more for coverage next.

File: msprime/provenance.py [hunk 1]
```diff
@@ -68,3 +76,108 @@ def get_environment():
     if _environment is None:
         _environment = _get_environment()
     return _environment
+
+
+class ProvenanceEncoderDecoder(json.JSONEncoder):
+    """
+      Extension of the `json` encoder that serializes arbitrary python class objects
```
Comment (jeromekelleher) [inline msprime/provenance.py]: Nitpick: we normally indent these like
```python
def something():
    """
    A docstring describing something
    """
```
happy to change this convention if you like (but it's a lot of code to update!)

File: tests/test_provenance.py [hunk 2]
```diff
@@ -77,6 +80,14 @@ def decode(self, prov):
         ns = builder.build_classes()
         return ns.TskitProvenance.from_json(prov)
 
+    def verify_equal(self, ts1, ts2):
+        ts1_tables = ts1.dump_tables()
+        ts2_tables = ts2.dump_tables()
+        # We exclude provenance here due to timestamps
```
Comment (jeromekelleher) [inline tests/test_provenance.py]: I prefer 
```python
ts1_tables.provenances.clear()
ts2_tables.provenances.clear()
self.assertEqual(ts1_tables, ts2_tables)
```
since this means that we don't need to know which tables are present (and maybe we'll add some more to tskit some day).

ts.dump_tables() will always return a copy of the tables so it's safe to do (ts.tables might return a read-only view on the underlying tables someday)

File: tests/test_provenance.py [hunk 3]
```diff
@@ -86,13 +97,100 @@ def test_simulation(self):
...
+            return t + 0.01
+
+        def bad_func(t):
+            if msprime:
+                pass
+            return t
+
+        with self.assertLogs('msprime.provenance', level='WARNING') as cm:
+            ts = msprime.simulate(random_seed=seeds[0],
+                                  population_configurations=pop_configs,
```
Comment (jeromekelleher) [inline tests/test_provenance.py]: I'm not a huge fan of visual indents, mainly because I think it makes the code layout more brittle. For example, here if I changed the variable name to ts1, I'd have to reindent this whole block. This exerts a pressure to *not* change the names of things, which means that code gets refactored less over time. Hanging indents are much more robust to code changes, so I'd do it like
```python
ts = msprime.simulate(
    random_seed=x, 
    population_configurations=pop_configs, 
   ....)
```

I agree it's a bit less pretty and not quite as easy for the eye to parse the structure though. Let's talk it through if you have strong opinions on indentation!

**Comment by jeromekelleher** [comment]
Ah yes, I did have some input beyond niggling nitpicks! Is there a perf cost here over what we were already doing @benjeffery? There's two scenarios I have in mind:
```python
import msprime

for ts in msprime.simulate(10, num_replicates=10**7):
    pass
```
Is there an effect now of capturing the provenance?

Also, what about when we run a large genetic map? This is the easiest way to do it: 
```
$ stdpopsim HomSap -g HapMapIIGRCh37 -c chr22 -o tmp.trees
```
are we spending much time making JSON? Is the tree sequence file much bigger because of the provenance? (This doesn't matter so much, as we'll update stdpopsim to use ``record_provenance=False``)

I'm not sure what's going to happen with from_ts. It seems likely that provenance will be pretty large there, and we'll need to warn users about this. Let's not worry about that for now though.

File: msprime/simulations.py [hunk 4]
```diff
@@ -460,13 +465,23 @@ def simulate(
         underlying object may be used for every TreeSequence
         returned which will most likely lead to unexpected behaviour.
     """
+    argspec = inspect.getargvalues(inspect.currentframe())
```
Comment (jeromekelleher) [inline msprime/simulations.py]: In the zillions of small replicates case, is it worthwhile skipping this step when record_provenance is False?

File: msprime/provenance.py (hunk 1 above)
Comment (benjeffery) [reply msprime/provenance.py]: I prefer your way as it gives more room on the line - just have to change my editor setup!

Have you thought of using https://pypi.org/project/flake8-docstrings/ to enforce aspects of docstrings like this?

File: tests/test_provenance.py (hunk 3 above)
Comment (benjeffery) [reply tests/test_provenance.py]: I usually let the editor reformat the code, so hadn't thought of that. However it does make me realise that diffs are cleaner if the indent is not visual as the change is limited to only the relevant lines. I'll configure my editor to this style and look into getting the linter to enforce it.

File: msprime/simulations.py (hunk 4 above)
Comment (benjeffery) [reply msprime/simulations.py]: I think this only gets called once for each call to `simulate`, not per replicate - but you're right there is no point calling it. Will move.

File: tests/test_provenance.py (hunk 2 above)
Comment (benjeffery) [reply tests/test_provenance.py]: :+1:

File: msprime/provenance.py (hunk 1 above)
Comment (jeromekelleher) [reply msprime/provenance.py]: > Have you thought of using https://pypi.org/project/flake8-docstrings/ to enforce aspects of docstrings like this?

No, but we definitely should! (Across the board for tskit projects too, this would be great to have).

File: tests/test_provenance.py (hunk 3 above)
Comment (jeromekelleher) [reply tests/test_provenance.py]: Great, thanks. I'm not sure about getting the linter to enforce this though --- sometimes visual indents do work well and some regular contributors prefer it. We're already quite prescriptive about style and I don't want to push people away. Having said that, I would kind of like to just use [black](https://github.com/psf/black) and be done with it. Style discussions can become very unproductive...

**Comment by benjeffery** [comment]
> Ah yes, I did have some input beyond niggling nitpicks! Is there a perf cost here over what we were already doing @benjeffery? There's two scenarios I have in mind:

For the simple case:
```python
for ts in msprime.simulate(10, num_replicates=10**5): 
    pass 
```
(You must have some sort of super-computer, 10**7 takes about 20min on my machine!)

master: `11.7 s ± 78.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`
This PR: `11.8 s ± 31 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`

I think this is expected as the only cost here is a tiny bit more JSON in each provenance entry as the parameters are simple. 

> Also, what about when we run a large genetic map? This is the easiest way to do it:
> $ stdpopsim HomSap -g HapMapIIGRCh37 -c chr22 -o tmp.trees
> are we spending much time making JSON? Is the tree sequence file much bigger because of the provenance? (This doesn't matter so much, as we'll update stdpopsim to use `record_provenance=False`)

I used `stdpopsim HomSap -g HapMapII_GRCh37 -c chr22 -o tmp.trees 2000`
The resulting file is 37MB for master, 39MB for this PR. This tallies up with the JSON length which was measured at 1,290,049 chars.  ~49ms were spent encoding this JSON.

I haven't looked into how provenance is stored but the combination of ~1MB provenances and large numbers or replicates is problematic.

> I'm not sure what's going to happen with from_ts. It seems likely that provenance will be pretty large there, and we'll need to warn users about this. Let's not worry about that for now though.

One option is to detect a very large provenance as it is generated (or detect provenance size * # of replicates), spit out a warning or even drop the provenance.

**Comment by jeromekelleher** [comment]
> One option is to detect a very large provenance as it is generated (or detect provenance size * # of replicates), spit out a warning or even drop the provenance.

That's a really good idea. I think if we detect a serialised provenance > 100K (say) we can raise a warning, telling the user that they should consider turning off provenance recording. This covers all bases nicely I think.

**Comment by grahamgower** [comment]
Has there been any consideration given to having a container file/format/convention for a collection of tree sequence files? Then one full provenance table could be stored for n replicates, plus the n seeds (or other unique information).

**Comment by jeromekelleher** [comment]
> Has there been any consideration given to having a container file/format/convention for a collection of tree sequence files? Then one full provenance table could be stored for n replicates, plus the n seeds (or other unique information).

No, not really. We've juggled using the idea for multiple chromosomes, but ended up deciding it wasn't worth the hassle. It seems unlikely many people are actually going to need to store lots of replicates, and if they do, that the provenance is going to be much of a burden.

There's no technical reason we couldn't though, we could imagine doing something like adding the tree sequence index as a prefix to all the column names in the underlying kastore as a way to do it. It'd be a **lot** of plumbing to update though.

**Comment by jeromekelleher** [comment]
Looks like something funny's happened with your git IDs @benjeffery? Unless there really is two of you...

**Comment by benjeffery** [comment]
> Looks like something funny's happened with your git IDs @benjeffery? Unless there really is two of you...

My evil twin strikes again! Set up the new machine with my BDI email, it's now added to my github account so should be all fixed.

**Comment by benjeffery** [comment]
Assuming the CI is now happy (in progress) I think this PR is complete.

There is one aspect of testing that i want to flag up. It is hard to test that the subclasses of e.g. `DemographicEvent` are giving the correct return for `asdict`, as in the general case a class could have modified and then stored an argument, meaning the modified version gets stored in provenance. One solution to this would to implement `asdict` for every subclass even if it just calls the method on the superclass, then in the tests iterate over every subclass to check that it has this implementation. This seemed like overkill for a small concern.

I've set the provenance limit to 2MB total over all replicates, this seemed to cover most advanced single tree sequence use cases. Happy to change if anyone has a strong opinion.

File: msprime/provenance.py [hunk 5]
```diff
@@ -20,6 +20,13 @@
 Common provenance methods used to determine the state and versions
 of various dependencies and the OS.
 """
+import base64
+import importlib
+import json
+import marshal
+import logging
+import types
```
Comment (jeromekelleher) [inline msprime/provenance.py]: Minor style point here. I like to split imports up into three groups: standard library, external imports and local imports. So, here it would be 
```python
import base64
￼import importlib
￼import json
￼import marshal
￼import logging
￼import types

import numpy
￼import tskit

import _msprime
```
This is a useful convention I think, as it lets you see in a glance what the module depends on.

File: msprime/provenance.py [hunk 6]
```diff
@@ -68,3 +76,131 @@ def get_environment():
...
+        size /= 1024.0
+    return f"{size:.{decimal_places}f}{unit}"
+
+
+def json_encode_provenance(provenance_dict, num_replicates=1):
+    """
+        Return a JSON representation of the provenance
+    """
+    prov = ProvenanceEncoderDecoder().encode(provenance_dict)
+    if len(prov) * num_replicates > 2_097_152:
```
Comment (jeromekelleher) [inline msprime/provenance.py]: I don't think we need to worry about the number of replicates. If someone does 10**6 replicates, they're almost certainly not storing the actual files, so it won't matter (and this warning will worry users who haven't seen it before).

I think we just need to worry about the case in which a single rep goes over the threshold. (2M seems as good as any)

File: msprime/simulations.py [hunk 7]
```diff
@@ -157,14 +158,23 @@ def _replicate_generator(
...
+    placeholder = '@@_MSPRIME_REPLICATE_INDEX_@@'
+    if provenance_dict is not None:
+        provenance_dict['parameters']['replicate_index'] = placeholder
+        encoded_provenance = provenance.json_encode_provenance(
+            provenance_dict, num_replicates)
+
     for j in range(num_replicates):
         sim.run(end_time)
-        tree_sequence = sim.get_tree_sequence(mutation_generator, provenance_record)
+        tree_sequence = sim.get_tree_sequence(
```
Comment (jeromekelleher) [inline msprime/simulations.py]: I'm not following the logic here - do we ever use the string "Provenance was disabled"? Can we pass ``None`` if not? I found myself having to think about what this does a bit too much, how about:
```python
provenance_record = None
if encoded_provenance is not None:
       provenance_record = encoded_provenance.replace(f'"{placeholder}"', str(j))
```
It's a bit more verbose, but clearer? I guess a rule of thumb for conditional assignments is "don't use them if you have to split lines"?

File: tests/test_provenance.py [hunk 8]
```diff
@@ -124,3 +215,87 @@ def test_mutate_numpy(self):
...
+        pop_configs = [msprime.PopulationConfiguration(5) for _ in range(2)]
+        ts = msprime.simulate(population_configurations=pop_configs,
+                              migration_matrix=[[0, 1], [1, 0]],
+                              demographic_events=[msprime.SimulationModelChange(
+                                  time=lambda t: t+0.1),
+                              ])
+        command, parsed_prov = msprime.provenance.parse_provenance(ts, None)
+        recreated = getattr(msprime, command)(**parsed_prov)
+        self.verify_equal(ts, recreated)
+
```
Comment (jeromekelleher) [inline tests/test_provenance.py]: Nice to split these up into discrete tests. First one is ``test_population_structure``, second is ``test_pedigree``, etc. Also nice to check a few more simple cases, like
```
def verify(ts):
    command, parsed_prov = msprime.provenance.parse_provenance(ts, None)
    recreated = getattr(msprime, command)(**parsed_prov)
    self.verify_equal(ts, recreated)


def test_sample_size(self):
    ts = msprime.simulate(10)
    self.verify(ts)

def test_random_seed(self):
     ts = msprime.simulate(10, random_seed=1234)
     self.verify(ts)
```

No need to be exhaustive, but good to capture a few of common usages here.

Also, good to test if we can round-round-trip --- we're not checking the provenance of the tree sequences we're generating from the parsed provenance. Can do this as a separate test I think.

File: tests/test_provenance.py [hunk 9]
```diff
@@ -124,3 +215,87 @@ def test_mutate_numpy(self):
...
+        ts = msprime.simulate(10, random_seed=5)
+        from_ts = tsutil.decapitate(ts, ts.num_edges // 2)
+        start_time = from_ts.tables.nodes.time.max()
+        final_ts = msprime.simulate(
+            from_ts=from_ts, start_time=start_time, random_seed=2)
+        command, parsed_prov = msprime.provenance.parse_provenance(final_ts, None)
+        recreated = getattr(msprime, command)(**parsed_prov)
+        self.verify_equal(final_ts, recreated)
+
+    def test_mutate_round_trip(self):
```
Comment (jeromekelleher) [inline tests/test_provenance.py]: Maybe better do this as it's own class? It could inherit the verify methods from a common superclass of TestSimulateRoundTrip.

File: tests/test_provenance.py [hunk 10]
```diff
@@ -124,3 +215,87 @@ def test_mutate_numpy(self):
...
+        command, parsed_prov = msprime.provenance.parse_provenance(ts, None)
+        recreated = getattr(msprime, command)(**parsed_prov)
+        self.verify_equal(ts, recreated)
+
+        ts = msprime.simulate(10, random_seed=5)
+        from_ts = tsutil.decapitate(ts, ts.num_edges // 2)
+        start_time = from_ts.tables.nodes.time.max()
+        final_ts = msprime.simulate(
+            from_ts=from_ts, start_time=start_time, random_seed=2)
+        command, parsed_prov = msprime.provenance.parse_provenance(final_ts, None)
```
Comment (jeromekelleher) [inline tests/test_provenance.py]: No point in specifying the provenance=None here, is there?

File: tests/test_provenance.py [hunk 11]
```diff
@@ -124,3 +215,87 @@ def test_mutate_numpy(self):
...
+        # Most testing of parsing is in the round tripping below
+        with self.assertRaises(ValueError):
+            ts = msprime.simulate(5)
+            prov = json.loads(ts.provenance(0).record)
+            prov['software']['name'] = 'skynet'
+            tables = ts.tables
+            tables.provenances.clear()
+            tables.provenances.add_row(json.dumps(prov))
+            command, prov = msprime.provenance.parse_provenance(tables.tree_sequence())
+
```
Comment (jeromekelleher) [inline tests/test_provenance.py]: We could use a test of the ``index`` argument here. Another test function would be most appropriate.

File: tests/test_provenance.py [hunk 12]
```diff
@@ -124,3 +215,87 @@ def test_mutate_numpy(self):
         self.assertEqual(decoded.parameters.start_time, 0)
         self.assertEqual(decoded.parameters.end_time, 100)
         self.assertEqual(decoded.parameters.keep, False)
+
+
+class TestParseProvenance(unittest.TestCase):
+    def test_parse_provenance(self):
```
Comment (jeromekelleher) [inline tests/test_provenance.py]: This is really ``test_parse_non_msprime`` I guess.

File: tests/test_provenance.py [hunk 13]
```diff
@@ -86,13 +90,95 @@ def test_simulation(self):
...
+                             " ('msprime',) which are not currently serialized"},
+                'model': None, '__class__': 'msprime.simulations.SimulationModelChange'}
+        ),
+        self.assertDictEqual(
+            parameters.demographic_events[2],
+            {'time': {
+                '__error__': 'Configuration function <lambda> refers to outer scope'
+                             ' variables which are not currently serialized'},
+                'model': None, '__class__': 'msprime.simulations.SimulationModelChange'})
+
```
Comment (jeromekelleher) [inline tests/test_provenance.py]: This is probably a different test case I guess?

File: msprime/provenance.py (hunk 5 above)
Comment (benjeffery) [reply msprime/provenance.py]: Cool, we should use https://github.com/timothycrosley/isort to do this as part of https://github.com/tskit-dev/tskit/issues/470 (which I'm assuming once we're happy with we'll bring across to the other repos?)

File: msprime/provenance.py (hunk 6 above)
Comment (benjeffery) [reply msprime/provenance.py]: I'm not so sure about this - even if they are not storing the files, it's still using up RAM. A simple, standard provenance is ~1500 chars, so the 10**6 replicate case would result in 1.5GB of provenance data in RAM without warning. 

I do get that anyone that uses a large number of replicates will always see this message though, which is not ideal. I propose we have 2M limit for single replicate and 500M total, and make the warning less scary by adding: "This is nothing to worry about unless you wish to save memory/storage"

File: msprime/simulations.py (hunk 7 above)
Comment (benjeffery) [reply msprime/simulations.py]: I agree the logic could be clearer. 
The intent here was that upon opening a file and looking at the provenance you clearly saw that it was disabled, rather than just getting "None", but I'm in danger of overthinking! I'll put it back to None.

File: tests/test_provenance.py (hunk 8 above)
Comment (benjeffery) [reply tests/test_provenance.py]: Nice catch - the round-round-trip found an edge case with functional arguments.

File: tests/test_provenance.py (hunk 9 above)
Comment (benjeffery) [reply tests/test_provenance.py]: Yep, cleaner.

File: tests/test_provenance.py (hunk 10 above)
Comment (benjeffery) [reply tests/test_provenance.py]: Yeah, that line is gone now after factoring out `verify`.

File: tests/test_provenance.py (hunk 11 above)
Comment (benjeffery) [reply tests/test_provenance.py]: Cool.

**Comment by benjeffery** [comment]
Thanks for the review @jeromekelleher! I've addressed all the points - just the warning message that I had some more thought on above. 

One aspect that came out of the test refactor was provenance in the `from_ts` case. Previously the provenance of the `from_ts` tree sequence was pre-pended to the output tree sequence. This didn't seem to make sense if we are storing the tree sequence in provenance, so I changed it to clear those provenances.

File: msprime/provenance.py (hunk 5 above)
Comment (jeromekelleher) [reply msprime/provenance.py]: Great idea! Yes, I'm imagining us getting the tskit repo fully up to date with new linting stuff before porting across to msprime, kastore, tsinfer etc.

File: msprime/provenance.py (hunk 6 above)
Comment (jeromekelleher) [reply msprime/provenance.py]: The standard approach for doing lots of replicates is 
```python
for ts in msprime.simulate(10, num_replicates=10**6):
    # do something with ts
```
So, only one tree sequence (and therefore provenance) is in memory at any one time. Unless you store the replicates in a list (which is a terrible idea anyway) the product of num_reps x provenance size isn't going to matter.

Unless I'm missing something here?

**Comment by jeromekelleher** [comment]
LGTM, thanks @benjeffery! I think there's just the issue about the size across replicates to resolve and we're good to go.

File: msprime/provenance.py (hunk 6 above)
Comment (benjeffery) [reply msprime/provenance.py]: picard_facepalm.gif 
Of course!

**Comment by benjeffery** [comment]
Ok, I've fixed my howling error! Should be good to go.

**Comment by jeromekelleher** [comment]
Excellent! OK, let's squash it back to 3 or 4 commits (the last few seem like they don't need to be in the history) and merge.

**Comment by benjeffery** [comment]
Done! When I was going through the history it boiled down to two distinct changes - adding the provenance mechanisms, and the new `replicate_index` argument to `simulate`.

**Comment by jeromekelleher** [comment]
Great stuff, thanks @benjeffery!

**Comment by benjeffery** [comment]
:tada: Thanks for all the great comments and help.
